### PR TITLE
Attempts to ensure raft transport timeout test achieves desired timeout.

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

Intermittent failures of WorkerSuite.TestTransportTimeout due to occasionally successfully dialling inside 1 millisecond.

Change uses a 1 nanosecond duration and a dummy listener that always causes the dial timeout.

## QA steps

Red/green changes to the test in question. The timeout value itself can be wound out and still gives the desired behaviour

## Documentation changes

No.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1749081

### CODE REVIEW

File: worker/raft/rafttransport/worker_test.go [hunk 1]
```diff
@@ -25,6 +25,7 @@ import (
 	coretesting "github.com/juju/juju/testing"
 	"github.com/juju/juju/worker/raft/rafttransport"
 	"github.com/juju/juju/worker/workertest"
+	"strings"
```
Comment (jameinel) [inline worker/raft/rafttransport/worker_test.go]: this ended up in the wrong import group

File: worker/raft/rafttransport/worker_test.go [hunk 2]
```diff
@@ -230,12 +231,28 @@ func (s *WorkerSuite) TestTransportWorkerStopped(c *gc.C) {
 
 func (s *WorkerSuite) TestTransportTimeout(c *gc.C) {
 	config := s.config
-	config.Timeout = time.Millisecond
+	config.Timeout = time.Nanosecond
```
Comment (jameinel) [inline worker/raft/rafttransport/worker_test.go]: so I'm a little concern that any sort of timeout like this is a chance for our side to be paused, and the other side to actually get the work done before the timeout gets noticed.
(Most sleep sort of things are 'sleep at least this long', I certainly don't think they can guarantee delivery time.)

If it was matched times "we set a deadline of 1ms, and on the other side we force a 10ms sleep", then I would be happier about it.

File: worker/raft/rafttransport/worker_test.go [hunk 3]
```diff
@@ -230,12 +231,28 @@ func (s *WorkerSuite) TestTransportWorkerStopped(c *gc.C) {
...
+		_, err := s.requestVote(worker)
+		return err
+	}
+
+	// This test previously had intermittent failures with a worker timeout
+	// set to 1 millisecond. With the nanosecond timeout, this should always
+	// break out first time, but it remains as a fail-safe.
+	deadline := time.After(coretesting.LongWait)
+	var err error
+	for err = rv(); !strings.Contains(err.Error(), "dial failed"); err = rv() {
```
Comment (jameinel) [inline worker/raft/rafttransport/worker_test.go]: I"m pretty sure (err == nil) will panic if you try to call err.Error().
Though if I set the timeout to 1s, then I get:
read tcp 127.0.0.1:59120->127.0.0.1:33506: i/o timeout

So presumably we always get an error. However, we should definitely be careful about getting a 'nil' back if we ever change how the worker acts.

I don't think you need to explain how the test 'used to work' in the comment. But having a good explanation of now is good.

// Watch out for an intermittent failure where the dial succeeds before we manage to fail the dial.

**Comment by manadart** [comment]
I have pushed the recommended fixes for the import group, non-nil error guard and comment.

As of CoB today however, I have still not managed to contrive a fail-safe by way of a shim with our own listener.

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**